src/features/courses/course-controller.ts

```diff
@@ -1050,6 +1157,491 @@ class CourseController {
+            logger.debug(`Regrading ${workbooks.length} attempts`);
+
+            question = question ?? await studentGrade.getQuestion({
+                where: {
+                    active: true
+                }
+            });
+            topic = topic ?? await question.getTopic({
+                where: {
+                    active: true
+                }
+            });
+
+            if (studentGrade.courseWWTopicQuestionId !== question.id) {
+                throw new IllegalArgumentException('studentGrade question id does not match the question\'s id');
+            }
+
+            if (question.courseTopicContentId !== topic.id) {
+                throw new IllegalArgumentException('question topic id does not match the topic\'s id');
+            }
+            
+            const solutionDate = moment(topic.deadDate).add(Constants.Course.SHOW_SOLUTIONS_DELAY_IN_DAYS, 'days');
+
+            // reset student grade before submitting
+            studentGrade.bestScore = 0;
+            studentGrade.overallBestScore = 0;
+            studentGrade.partialCreditBestScore = 0;
+            studentGrade.effectiveScore = 0;
+            studentGrade.legalScore = 0;
+            studentGrade.numAttempts = 0;
+            studentGrade.numLegalAttempts = 0;
+            studentGrade.numExtendedAttempts = 0;
+            studentGrade.lastInfluencingAttemptId = null;
+            studentGrade.lastInfluencingCreditedAttemptId = null;
+            studentGrade.lastInfluencingLegalAttemptId = null;
+            studentGrade.firstAttempts = 0;
+            studentGrade.latestAttempts = 0;
+            
+            const gradeOverrides: StudentGradeOverride[] = await studentGrade.getOverrides({
+                order: ['id'],
+                where: {
+                    active: true
+                }
+            });
+            
+            const workbooksAndOverrides: (StudentWorkbook | StudentGradeOverride)[] = [...workbooks, ...gradeOverrides];
+            const sortedWorkbooksAndOverrides = workbooksAndOverrides.sort((first: StudentWorkbook | StudentGradeOverride, second: StudentWorkbook | StudentGradeOverride): number => {
+                const getDate = (object: StudentWorkbook | StudentGradeOverride): Date => {
+                    if (object instanceof StudentWorkbook) {
+                        return object.time;
+                    } else {
+                        if (!(object instanceof StudentGradeOverride)) {
+                            logger.error('Invalid type, should be StudentWorkbook | StudentGradeOverride');
+                        }
+                        // could always be createdAt...
+                        // time for workbook is redundant and should always be the same as created at
+                        // That being said
+                        return object.createdAt;
+                    }
+                };
+                return getDate(first).getTime() - getDate(second).getTime();
```

Comment:
Does this take/need to take into account the date? I've seen getTime in a couple of places.